src/cosmos/CosmosEngine.ts

```diff
@@ -914,12 +915,84 @@ export class CosmosEngine extends CurrencyEngine<
     return edgeTransaction
   }
 
-  async signMessage(message: string, privateKeys: JsonObject): Promise<string> {
+  async signMessage(
+    message: string,
+    privateKeys: JsonObject,
+    opts?: JsonObject
+  ): Promise<string> {
     const keys = asCosmosPrivateKeys(this.currencyInfo.pluginId)(privateKeys)
+    const signer = await this.tools.createSigner(keys.mnemonic)
+
+    if (opts?.otherParams != null) {
+      switch (opts.otherParams.method) {
+        case 'cosmos_getAccounts': {
+          asCosmosWcGetAccountsRpcPayload(opts.otherParams)
+          const accounts = await signer.getAccounts()
+          const result = accounts.map(account => {
+            return {
+              ...account,
+              pubkey: base64.stringify(account.pubkey)
+            }
+          })
+          return JSON.stringify(result)
+        }
+        case 'cosmos_signDirect': {
+          const payload = asCosmosWcSignDirectRpcPayload(opts.otherParams)
+          const {
+            params: { signDoc }
+          } = payload
+          const signDirectDoc = SignDoc.fromPartial({
+            accountNumber: longify(signDoc.accountNumber),
+            authInfoBytes: safeParse(signDoc.authInfoBytes),
+            bodyBytes: safeParse(signDoc.bodyBytes),
+            chainId: signDoc.chainId
+          })
+          const signResponse = await signer.signDirect(
+            this.walletInfo.keys.bech32Address,
+            signDirectDoc
+          )
+          const result = {
+            signature: signResponse.signature,
+            signed: payload.params.signDoc
+          }
+          return JSON.stringify(result)
+        }
+        case 'cosmos_signAmino': {
+          const payload = asCosmosWcSignAminoRpcPayload(opts.otherParams)
+          const {
+            params: { signDoc }
+          } = payload
```

Comment:
I'd prefer to see `const { signDoc } = payload.params`, since dots are easier to read than curly braces.